### PR TITLE
Extract PackageTree and stuff into subpackage

### PR DESCRIPTION
**Pull request by narqo**
I've started working on #166 to get some overview of the project's internals. While the process hasn't finished yet, I feel it might be a good time to get some early feedback.

Tests for `solve_bimodal_test.go` are failing as they rely on private `workmap` of pkgtree. I feel we could refactor this piece, but I'm not sure what is the best way to do it. Definitely, need to dive more deeply into the project's code base.

P. S. It looks like `fs` package might be a candidate for having `fastWalk` function from #180, doesn't it?

### CODE REVIEW

File: analysis.go [hunk 1]
```diff
@@ -30,943 +20,3 @@ func init() {
 	archListString := "386 amd64 amd64p32 arm armbe arm64 arm64be ppc64 ppc64le mips mipsle mips64 mips64le mips64p32 mips64p32le ppc s390 s390x sparc sparc64"
 	archList = strings.Split(archListString, " ")
```
Comment (narqo) [inline analysis.go]: @sdboyer 

I haven't found any references to these variables. Could you provide some details, why they are here?

File: pkgtree/pkgtree.go [hunk 2]
```diff
@@ -0,0 +1,898 @@
...
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+	"unicode"
+)
+
+// Stored as a var so that tests can swap it out. Ugh globals, ugh.
+var IsStdLib = doIsStdLib
```
Comment (sdboyer) [inline pkgtree/pkgtree.go]: This is one of those things I'd really strongly prefer not to export, but we have the call to it in the solver.

If we can't think of any clever solutions, then I'd prefer we duplicate the function in both packages than have it exposed. The only solution I can think of is adding a stdlib filtering param to `ToReachMap()`...but that doesn't take care of the calls in `rootdata.go`.

**Comment by sdboyer** [comment]
woohoo more contributors! welcome! 👏 👏 🎉 

I think the best route with the `wmToReach()` call in `computeBimodalExternalMap()` is to just bite the bullet and fake up a `PackageTree` that'll produce the same `ReachMap` when `ToReachMap()` is called on it as we currently get from calling `wmToReach()`. I have a nagging memory of some other plan for refactoring it, but it's been a while, and that seems sanest when I look back at it now.

> P. S. It looks like `fs` package might be a candidate for having `fastWalk` function from #180, doesn't it?

Let's lump it in with the `pkgtree` package for now, unexported. We can consider factoring it out later, but I don't want to expose it, because then we have to support it/can't change it.

File: analysis.go (hunk 1 above)
Comment (sdboyer) [inline analysis.go]: Ah, yeah. I thought I was going to get to this earlier, but I still haven't.

Basically, they're supposed to be for this: golang/dep#291

File: solve_test.go [hunk 3]
```diff
@@ -47,6 +49,8 @@ func overrideIsStdLib() {
 	isStdLib = func(path string) bool {
 		return false
 	}
+	// NOTE(narqo): this is an ugly hack! One have to think about better way to do cross-package mocking.
+	pkgtree.MockIsStdLib()
```
Comment (narqo) [inline solve_test.go]: I really don't like this, but currently, I don't see another way to make fixtures work.

**Comment by sdboyer** [comment]
(btw, i'd prefer that you just push up commits as you go. i realize i might be weird in this regard, but it's easier for me to keep track of a running diff)

**Comment by narqo** [comment]
> [..] i'd prefer that you just push up commits as you go [..]

No problem. Will try to do some meaningful commits, not some random lazy "wip!"s. :)

I think I'm done with #166 and waiting for your review.

**Comment by sdboyer** [comment]
OK, great. Might take me a couple days, but I'll definitely try to get this to happen before moving gps into dep (golang/dep#300).

File: solve_test.go (hunk 3 above)
Comment (sdboyer) [inline solve_test.go]: Yknow...I think maybe we could just expose `var IsStdlib` on the subpackage, then have tests overwrite it in much the way they do now. As long as there's a doc on it indicating its intended use, and that overwriting it except for test purposes will cause undefined behavior, then I'm not too bothered.

File: solve_bimodal_test.go [hunk 4]
```diff
@@ -1153,46 +1155,47 @@ func (sm *bmSourceManager) GetManifestAndLock(id ProjectIdentifier, v Version) (
...
 			}
-			workmap[pkg.path] = w
 		}
-
-		reachmap, em := wmToReach(workmap, true)
+		reachmap, em := ptree.ToReachMap(false, true, true, nil)
 		if len(em) > 0 {
 			panic(fmt.Sprintf("pkgs with errors in reachmap processing: %s", em))
 		}
+		fmt.Printf("reachmap: %+v\n", reachmap)
```
Comment (sdboyer) [inline solve_bimodal_test.go]: Left over from debugging?

File: analysis.go [hunk 5]
```diff
@@ -1,17 +1,7 @@
 package gps
```
Comment (sdboyer) [inline analysis.go]: let's just get rid of this file entirely - move it into solver.go.

File: fs/fs.go [hunk 6]
```diff
@@ -1,7 +1,7 @@
-package gps
+package fs
```
Comment (sdboyer) [inline fs/fs.go]: I think I'd rather make this `internal/fs`, at least for now - these funcs aren't really intended to be a part of the public surface area.

**Comment by narqo** [comment]
Ready for next iteration.

**Comment by sdboyer** [comment]
(sorry for delay - very very busy week, changing jobs et. all)

Looks like the change you made to the coverage counter ended up just reporting the coverage from the pkgtree subdir :( Get that fixed, and we can merge!

Also, I decided on a much more pleasant way to deal with the `isStdLib()` func - we'll add a parameter to `Flatten()` that lets the caller inject a filter func - but can take care of that in a follow-up PR.

**Comment by narqo** [comment]
This is weird, AppVeyor's tests are failing with "nil pointer dereference" from `os/exec`. But it doesn't look like the changes are the reason. @sdboyer could you help me here?

**Comment by sdboyer** [comment]
That seems like it's an ephemeral error. (Which is itself not reassuring, but...)

**Comment by sdboyer** [comment]
Let's remove that latest revert (9e5eeb6), we do want those errors renamed. With any luck, tests will pass on re-run.

**Comment by narqo** [comment]
> Let's remove that latest revert (9e5eeb6), we do want those errors renamed

After spending some time digging through go's standard library internals, it seems that naming error types as `SomethingError` is actually a common practice. See https://talks.golang.org/2014/names.slide#15 If you don't mind, I'd like to consider that my initial renaming in the fist commit of the PR, was a mistake.

**Comment by sdboyer** [comment]
indeed! i'm glad that's the standard, i rather like that better :D

OK, i think i should be able to merge this tonight, then. just finishing
up a couple things elsewhere, first.

**Comment by sdboyer** [comment]
OK, pulling this in. Thanks for all your work and patience on this!

**Pull request merged**